File: javasolutions/p236/LowestCommonAncestor.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        if not root or root == p or root == q:
            return root

        leftRt = self.lowestCommonAncestor(root.left, p, q)
        rightRt = self.lowestCommonAncestor(root.right, p, q)

        if leftRt and rightRt:
            return root

        if leftRt:
            return leftRt

        if rightRt:
            return rightRt

        return None

File: programming_language/javasolutions/p236/LowestCommonAncestor.java
/**
 * https://leetcode.com/problems/lowest-common-ancestor-of-a-binary-tree/?tab=Description
 *
 * The lowest common ancestor is defined between two node v and w, as the lowest node in T that
 * has both v and w as descendants (where we allow a node to be a descendant of itself)
 */

package javasolutions.p236;

import javasolutions.TreeNode;

public class LowestCommonAncestor {
  public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
    if (root == null || root == p || root == q) return root;

    TreeNode left = lowestCommonAncestor(root.left, p, q);
    if (left != null && left != p && left != q) return left;

    TreeNode right = lowestCommonAncestor(root.right, p, q);
    if(right != null && right != p && right != q) return right;

    if(left == null) {
      return right;
    } else if (right == null) {
      return left;
    } else {
      return root;
    }
  }
}